src/garbage_collector.rs

```diff
@@ -0,0 +1,253 @@
+    manifest_store: Arc<ManifestStore>,
+    table_store: Arc<TableStore>,
+    options: GarbageCollectorOptions,
+    external_rx: crossbeam_channel::Receiver<GarbageCollectorMessage>,
+    db_stats: Arc<DbStats>,
+}
+
+impl GarbageCollectorOrchestrator {
+    /// Collect garbage from the manifest store. This will delete any manifests
+    /// that are older than the minimum age specified in the options.
+    async fn collect_garbage_manifests(&self) -> Result<(), SlateDBError> {
+        let min_age = self
+            .options
+            .manifest_options
+            .map_or(DEFAULT_MIN_AGE, |opts| opts.min_age);
+        let mut manifest_metadata_list = self.manifest_store.list_manifests(..).await?;
+
+        // Remove the last element so we never delete the latest manifest
+        manifest_metadata_list.pop();
+
+        // TODO Should exclude snapshotted manifests when we implement snapshots
+
+        // Delete manifests older than min_age
+        for manifest_metadata in manifest_metadata_list {
+            let min_age = chrono::Duration::from_std(min_age).expect("invalid duration");
+
+            if Utc::now().signed_duration_since(manifest_metadata.last_modified) > min_age {
+                if let Err(e) = self
+                    .manifest_store
+                    .delete_manifest(manifest_metadata.id)
+                    .await
+                {
+                    error!("Error deleting manifest: {}", e);
+                } else {
+                    self.db_stats.gc_manifest_count.inc();
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Collect garbage from the WAL SSTs. This will delete any WAL SSTs that are
+    /// older than the minimum age specified in the options and are also older than
+    /// the last compacted WAL SST.
+    async fn collect_garbage_wal_ssts(&self) -> Result<(), SlateDBError> {
+        let last_compacted_wal_sst_id = self
+            .manifest_store
+            // Get the latest manifest so we can get the last compacted id
+            .read_latest_manifest()
+            .await?
+            .ok_or_else(|| SlateDBError::ManifestMissing)?
+            // read_latest_manifest returns (id, manifest) but we only care about the manifest
+            .1
+            .core
+            .last_compacted_wal_sst_id;
+        let min_age = self
+            .options
+            .wal_options
+            .map_or(DEFAULT_MIN_AGE, |opts| opts.min_age);
+        let sst_ids_to_delete = self
+            .table_store
+            .list_wal_ssts(..last_compacted_wal_sst_id)
+            .await?
+            .into_iter()
+            .filter(|wal_sst| {
+                let min_age = chrono::Duration::from_std(min_age).expect("invalid duration");
+                Utc::now().signed_duration_since(wal_sst.last_modified) > min_age
+            })
+            .map(|wal_sst| wal_sst.id)
+            .collect::<Vec<_>>();
+
+        for id in sst_ids_to_delete {
```

Comment:
I think there is a bug in `fence_writers`, when garbage collection deletes all the wal files. Deleting all the wal can happen in couple of cases i) WAL Disabled, and it has been more than `min_age` since the last writer ii) WAL Enabled, but there hasn't been any `put` for `min_age`.

`fence_writers`  lists all the wal, and writes max + 1.
If there are no wal's it would write `1.sst`. 

`fence_writers` should consider `last_compacted_wal` as well. 

In addition, for garbage collection, shall we skip the last WAL?

Reply:
I think this is addressed in this part of my code:

```rust
        let sst_ids_to_delete = self
            .table_store
            .list_wal_ssts(..last_compacted_wal_sst_id)  // <-- This is an exclusive range, so the last WAL is not deleted
            .await?
            .into_iter()
            .filter(|wal_sst| {
                let min_age = chrono::Duration::from_std(min_age).expect("invalid duration");
                utc_now.signed_duration_since(wal_sst.last_modified) > min_age
            })
            .map(|wal_sst| wal_sst.id)
            .collect::<Vec<_>>();
```

I test this in the tests, too.

As for case (i), where the WAL is disabled, there are two situations:

- WAL has always been disabled
- WAL was used but has since been disabled

In the former case, there are no WAL files at all, so the GC won't have any effect on the `wal` directory.

In the later case, the `last_compacted_wal_sst_id` will be set, and GC will honor that.

Does this address make sense/address your concern?

Reply:
yes! I missed the exclusive range part.